include/libopencm3/stm32/g0/adc.h

```diff
@@ -65,8 +66,29 @@
 /** ADC_CALFACT Calibration factor register */
 #define ADC_CALFACT(adc)	MMIO32((adc) + 0xB4)
 
-/** ADC_OR Option register */
-#define ADC_OR(adc)			MMIO32((adc) + 0xD0)
+#define ADC1_ISR		ADC_ISR(ADC1)
+#define ADC1_IER		ADC_IER(ADC1)
+#define ADC1_CR			ADC_CR(ADC1)
+#define ADC1_CFGR1		ADC_CFGR1(ADC1)
+#define ADC1_CFGR2		ADC_CFGR2(ADC1)
+#define ADC1_SMPR1		ADC_SMPR1(ADC1)
+#define ADC1_TR1		ADC_TR1(ADC1)
+#define ADC1_DR			ADC_DR(ADC1)
+#define ADC1_CCR		ADC_CCR(ADC1)
+#define ADC1_CHSELR		ADC_CHSELR(ADC1)
+
+#define ADC1_AWD1TR		ADC_AWD1TR(ADC1)
+#define ADC1_AWD2TR		ADC_AWD2TR(ADC1)
+#define ADC1_AWD3TR		ADC_AWD3TR(ADC1)
+#define ADC1_AWD2CR		ADC_AWD2CR(ADC1)
+#define ADC1_AWD3CR		ADC_AWD3CR(ADC1)
+#define ADC1_CALFACT		ADC_CALFACT(ADC1)
```

Comment:
I was personally hoping to leave this style behind with the older parts.  Are you adding this to be more compatible with some other existing code?  I'd rather remove these explicit expanded forms than add more of them, it just makes SO MUCH boilerplate.

Still, even if you realllly want to, these would go in adc_common_v2.h (see how they're done in _v1.h)  but really, it just feels superfluous and I'd rather drop it from v1 than add it to v2....

Reply:
Personally, I'd say having "shortcut" macros is  very a good thing when instead of `ADC_CR(ADC1)` one can use `ADC_CR` (or at least `ADC1_CR`, ok, ok that's at least something) remove lots of excessive clutter when it's not needed. So I'd rather vote for these macros since they are already present in families like F0/F1. I'll move them to `adc_common_v2.h` as you advised.

Reply:
Hmm, not so fast.... Please take a look at the F3:
https://github.com/libopencm3/libopencm3/blob/master/include/libopencm3/stm32/f3/adc.h#L54

It defines aliases right there yet uses `adc_common_v2` - the same approach that is practised in all the headers that use `adc_common_v1`. In terms of consistency with the rest of the code, it's not easy to put the above aliases to `adc_common_v2` as you suggested because it would mean removing them from `f0/adc.h` and `f3/adc.h`. So why is this an issue now if it's **already** in `f0/adc.h` and `f3/adc.h`?

Reply:
Ok, let me try to convince you that these aliases should stay where they are now, i.e. in their individual `adc.h` files like it's already done for the F0 and F3 (see the link above).

> see how they're done in _v1.h

This is how they are done in V1: https://github.com/libopencm3/libopencm3/blob/master/include/libopencm3/stm32/common/adc_common_v1.h#L66-L153
You can see that there is one unconditional section and two _#ifdef_ sections in `adc_common_v1.h` covering aliases for three (3) ADCx_BASE defines in total. Additionally, all the original (aliased) registers are defined in the same file. Moreover, common registers are defined in `adc_common_v1_multi.h`, hence they don't need aliases. And finally, the ADC in the F1/F4/F7 (v1) is basically the same, i.e. the list of its registers is the same for all the families and is short.

Unlike adc_common_v1, adc_common_v2 is split into three files - `adc_common_v2.h`, `adc_common_v2_single.h` and `adc_common_v2_multi.h` - which all have their own instance-based ADC register definitions for certain variantions. The maximum number of ADCs is in the G4 - five (5) instances. Another (bigger) problem is that all the families that use `adc_common_v2*.h`, i.e. the F0/F3/G0/G4/L0/L4, all have differences. The register names are very different at times - some have one ADC_TR whereas others have several ADC_TRx or even ADC_AWDxTR instead, some have ADC_SMPR whereas others ADC_SMPR1/ADC_SMPR2, etc. These differences are covered in their individual `adc.h` headers. Hence, the best place for aliases is also in their `adc.h` files which is *already* the case for the F0 and F3. Gathering the complete list of registers scattered across so many files, taking into account all the differences in their names and quantities, multiplying that by 5 (maximum number of instances) and putting aliases in one spot is going to be a very painful and generally useless endeavour. I think that is why V2 aliases are currently defined in individual `adc.h` header files, and I'd vote for that 100%. Hope you'll agree.

Reply:
These expansions/instanciations for each peripheral of the same type feel useless to me, because the provided routines are providing an abstraction for all similar peripherals, and direct access to specific peripherals is unlikely to be needed in user code.

Reply:
Personally, I **always** use these aliases in my projects. They are very useful when you do bare-metal programming which I do. Fortunately, they are **already** present in many peripherals including timers, ADC, USARTs, etc. I'm simply trying to add consistency across the F0/G0/L0 family. I'm quite surprised as to why there's so much resistance against it.

Reply:
I think the resistance is related to abstraction, if you want relatively portable code, you should be using the API functions instead of direct register accesses in your app.

I understand that you could have specific behaviours in your app. In that case, the shorthands are a time saver.

But ultimately you could wrap these into useful functions and submit them to libopencm3. Then, they can be implemented for other families.

Reply:
Adding abstract API functions to libopencm3 is beyond the purpose of this PR. It's solely dedicated to adding consistency across the F0/G0/L0 family.

Reply:
What about submitting it without these changes to give the PR more chance to be merged?